tests/teststun.nim

```diff
@@ -0,0 +1,130 @@
+import ../webrtc/stun/stun_message
+import ../webrtc/stun/stun_attributes
+import ../webrtc/stun/stun_utils
+import ./asyncunit
+
+proc newRng(): ref HmacDrbgContext =
+  HmacDrbgContext.new()
+
+proc usernameProvEmpty(): string = ""
+proc usernameProvTest(): string {.raises: [], gcsafe.} = "TestUsername"
+proc usernameCheckTrue(username: seq[byte]): bool {.raises: [], gcsafe.} = true
+proc usernameCheckFalse(username: seq[byte]): bool {.raises: [], gcsafe.} = false
+proc passwordProvEmpty(username: seq[byte]): seq[byte] {.raises: [], gcsafe.} = @[]
+proc passwordProvTest(username: seq[byte]): seq[byte] {.raises: [], gcsafe.} = @[1'u8, 2, 3, 4]
+
+suite "Stun message encoding/decoding":
+  test "Get BindingRequest + encode & decode with a set username":
+    var
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(
+        udpConn,
+        TransportAddress(AnyAddress),
+        iceControlling=true,
+        usernameProvider=usernameProvTest,
+        usernameChecker=usernameCheckTrue,
+        passwordProvider=passwordProvEmpty,
+        newRng()
+      )
+      msg = conn.getBindingRequest()
+      encoded = msg.encode(@[1'u8, 2, 3, 4])
+      decoded = StunMessage.decode(encoded)
+      messageIntegrity = decoded.attributes[^2]
+      fingerprint = decoded.attributes[^1]
+
+    decoded.attributes = decoded.attributes[0 ..< ^2]
+    check:
+      decoded == msg
+      messageIntegrity.attributeType == AttrMessageIntegrity.uint16
+      fingerprint.attributeType == AttrFingerprint.uint16
+    conn.close()
+
+  test "Get BindingResponse from BindingRequest + encode & decode":
+    var
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(
+        udpConn,
+        TransportAddress(AnyAddress),
+        iceControlling=false,
+        usernameProvider=usernameProvTest,
+        usernameChecker=usernameCheckTrue,
+        passwordProvider=passwordProvEmpty,
+        newRng()
+      )
+      bindingRequest = conn.getBindingRequest()
+      bindingResponse = conn.getBindingResponse(bindingRequest)
+      encoded = bindingResponse.encode(@[1'u8, 2, 3, 4])
+      decoded = StunMessage.decode(encoded)
+      messageIntegrity = decoded.attributes[^2]
+      fingerprint = decoded.attributes[^1]
+
+    decoded.attributes = decoded.attributes[0 ..< ^2]
+    check:
+      bindingResponse == decoded
+      messageIntegrity.attributeType == AttrMessageIntegrity.uint16
+      fingerprint.attributeType == AttrFingerprint.uint16
+
+suite "Stun checkForError":
+  test "checkForError: Missing MessageIntegrity or Username":
+    var
+      udpConn = UdpConn.init(AnyAddress)
+      conn = StunConn.init(
+        udpConn,
+        TransportAddress(AnyAddress),
+        iceControlling=false,
+        usernameProvider=usernameProvEmpty, # Use of an empty username provider
+        usernameChecker=usernameCheckTrue,
+        passwordProvider=passwordProvEmpty,
+        newRng()
+      )
+      bindingRequest = conn.getBindingRequest()
```

Comment:
Do we need a conn to create a request? It shouldn't be necessary for testing encoding/decoding. I see a connection as a medium for reading/writing, not creating msgs. Makes sense?

Reply:
Ok, the alternative would be to pass many parameters to this proc. In this case, it is sensible.